ipn/ipnlocal/local.go

```diff
@@ -4837,9 +4840,45 @@ func (b *LocalBackend) doSetHostinfoFilterServices() {
 	c := len(hi.Services)
 	hi.Services = append(hi.Services[:c:c], peerAPIServices...)
 	hi.PushDeviceToken = b.pushDeviceToken.Load()
+
+	if hostinfoPeerAPIPortsChanged(oldServices, hi.Services) {
```

Comment:
Nit: if you simply do this check on line 4831, I think you can skip using the `oldServices` variable and just compare `peerAPIServices` with `hi.Services`.